src/routes/movies.py

```
@@ -1,13 +1,71 @@
-
-from database import get_db, MovieModel
-from database.models import CountryModel, GenreModel, ActorModel, LanguageModel
+from fastapi import APIRouter, HTTPException, Request
+from pydantic import ValidationError
+from starlette.responses import JSONResponse
 
+from crud import get_movies, create_movie, get_movie_by_id, remove_movie, edit_movie
+from routes.dependencies import PaginationDep, SessionDep
+from schemas.movies import MovieListResponseSchema, MovieCreateSchema, MovieDetailSchema, MovieUpdateSchema
 
 router = APIRouter()
 
-# Write your code here
+
+@router.get("/movies/", response_model=MovieListResponseSchema)
+async def list_movies(
+        request: Request,
+        params: PaginationDep,
+        db: SessionDep
+):
+    movies = await get_movies(
+        request=request,
+        page=params["page"],
+        per_page=params["per_page"],
+        db=db
+    )
+    if not movies:
+        raise HTTPException(status_code=404, detail="No movies found.")
+    return movies
+
+
+@router.post("/movies/", status_code=201, response_model=MovieDetailSchema)
+async def add_movie(movie_data: dict, db: SessionDep):
+    try:
+        movie = MovieCreateSchema(**movie_data)
+    except ValidationError:
+        raise HTTPException(status_code=400, detail="Invalid movie data")
+
+    db_movie = await create_movie(movie=movie, db=db)
+    return db_movie
+
+
+@router.get("/movies/{movie_id}/", response_model=MovieDetailSchema)
+async def retrieve_movie(movie_id: int, db: SessionDep):
+    db_movie = await get_movie_by_id(movie_id=movie_id, db=db)
+    if not db_movie:
+        raise HTTPException(status_code=404, detail="Movie with the given ID was not found.")
+    return db_movie
+
+
+@router.delete("/movies/{movie_id}/")
+async def delete_movie(movie_id: int, db: SessionDep):
+    db_movie = await get_movie_by_id(movie_id=movie_id, db=db)
+    if not db_movie:
+        raise HTTPException(status_code=404, detail="Movie with the given ID was not found.")
```

The logic to retrieve a movie by its ID and raise an HTTPException if it's not found is duplicated across three different endpoints: `retrieve_movie`, `delete_movie`, and `update_movie`. To adhere to the DRY (Don't Repeat Yourself) principle, this common logic should be extracted into a shared dependency. A new dependency function could encapsulate this 'get or 404' behavior, cleaning up the route handlers and centralizing the logic for better maintainability.